server/src/main/java/io/crate/expression/scalar/NumNullTermsPerDocQuery.java

```diff
@@ -0,0 +1,225 @@
+    private final IntPredicate matches;
+
+    private static IntUnaryOperator getNumNullTermsPerDocFunction(LeafReader reader, Reference ref, Function<ColumnIdent, Reference> getRef) {
+        DataType<?> elementType = ArrayType.unnest(ref.valueType());
+        switch (elementType.id()) {
+            case BooleanType.ID:
+            case ByteType.ID:
+            case ShortType.ID:
+            case IntegerType.ID:
+            case LongType.ID:
+            case TimestampType.ID_WITH_TZ:
+            case TimestampType.ID_WITHOUT_TZ:
+            case FloatType.ID:
+            case DoubleType.ID:
+            case GeoPointType.ID:
+                return numValuesPerDocForSortedNumeric(reader, ref, getRef);
+
+            case NumericType.ID: {
+                NumericType numericType = (NumericType) elementType;
+                Integer precision = numericType.numericPrecision();
+                if (precision == null || precision > NumericStorage.COMPACT_PRECISION) {
+                    return numValuesPerDocForSortedSetDocValues(reader, ref, getRef);
+                }
+                return numValuesPerDocForSortedNumeric(reader, ref, getRef);
+            }
+
+            case StringType.ID:
+            case CharacterType.ID:
+            case BitStringType.ID:
+            case IpType.ID:
+                return numValuesPerDocForSortedSetDocValues(reader, ref, getRef);
+
+            default:
+                throw new UnsupportedOperationException("NYI: " + elementType);
+        }
+    }
+
+    private static IntUnaryOperator numValuesPerDocForSortedSetDocValues(LeafReader reader, Reference ref, Function<ColumnIdent, Reference> getRef) {
+        final SortedSetDocValues numNonNullTerms;
+        final SortedNumericDocValues numAllTerms;
+        try {
+            numNonNullTerms = DocValues.getSortedSet(reader, ref.storageIdent());
+            numAllTerms = DocValues.getSortedNumeric(reader, ArrayIndexer.toArrayLengthFieldName(ref, getRef));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return doc -> {
+            try {
+                return numAllTerms.advanceExact(doc) && numNonNullTerms.advanceExact(doc) ?
+                    Math.toIntExact(numAllTerms.nextValue()) - numNonNullTerms.docValueCount() : -1;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        };
+    }
+
+    private static IntUnaryOperator numValuesPerDocForSortedNumeric(LeafReader reader, Reference ref, Function<ColumnIdent, Reference> getRef) {
+        final SortedNumericDocValues numNonNullTerms;
+        final SortedNumericDocValues numAllTerms;
+        try {
+            numNonNullTerms = DocValues.getSortedNumeric(reader, ref.storageIdent());
+            numAllTerms = DocValues.getSortedNumeric(reader, ArrayIndexer.toArrayLengthFieldName(ref, getRef));
```

Comment:
`NumNullTermsPerDocQuery` is mostly identical to `NumTermsPerDocQuery` where the main difference are these lines.